src/server/api/routers/communication.ts

```diff
@@ -0,0 +1,138 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-return */
+import { observable } from "@trpc/server/observable";
+import { EventEmitter } from "events";
+import { z } from "zod";
+import { createTRPCRouter, protectedProcedure } from "~/server/api/trpc";
+
+const ee = new EventEmitter();
+
+export type Message = {
+  sessionId: string;
+  senderId: string;
+  senderName: string;
+  message: string;
+  createdAt: Date;
+};
+
+// Methods include userId as well because users do not have unique names, so who is typing is decided by the userId
+export const messagesRouter = createTRPCRouter({
+  getAllSessionMessages: protectedProcedure
+    .input(
+      z.object({
+        sessionId: z.string(),
+      }),
+    )
+    .query(async ({ ctx, input }) => {
+      const { sessionId } = input;
+
+      const messages = await ctx.prismaPostgres.sessionMessage.findMany({
+        where: {
+          sessionId,
+        },
+        orderBy: {
+          createdAt: "asc",
+        },
+      });
+
+      return messages;
+    }),
+
+  addMessage: protectedProcedure
+    .input(
+      z.object({
+        sessionId: z.string(),
+        senderId: z.string(),
+        senderName: z.string(),
+        message: z.string(),
+      }),
+    )
+    .mutation(async ({ ctx, input }) => {
+      const { sessionId, senderId, senderName, message } = input;
+
+      const messageObject = await ctx.prismaPostgres.sessionMessage
+        .create({
+          data: {
+            sessionId,
+            senderId,
+            senderName,
+            message,
+          },
+        })
+        .then((req) => {
+          return req;
+        });
+
+      ee.emit("message", messageObject);
+      ee.emit("typing", {
+        sessionId,
+        userId: senderId,
+        userName: senderName,
+        isTyping: false,
+      });
+
+      return messageObject;
+    }),
+
+  subscribeToSessionMessages: protectedProcedure
```

Comment:
Wouldn't it be better for subscription if we simplify to only one subscription for the communications route.
So when someone is typing we send a typing message, and when someone has sent a message, we send a message object, but we only use one subscription to do that

Reply:
Basicaly shift the ee.on("typing", ...) code here

Reply:
you mean one subscription emits two datatypes?

Reply:
One data type which includes an optional message and partner is typing

Reply:
done